docs/DdnRelocatableShaderElf.md

```diff
@@ -0,0 +1,209 @@
+shaders in a reasonable way.  For exmaple, the regiser count entries will take
+the maximum value from all of the unlinked shaders.  Some other values are
+bitfields
+where a bit is set if a feature is used.  These are merged by taking the
+bitwise-or of the values.  The code that merges the PAL metadata can be found in
+[PalMetadata.cpp](../lgc/state/PalMetadata.cpp)
+in the function `PalMetadata::mergeFromBlob`.
+
+## PAL metadata finalization
+
+PAL metadata finalization is done after linking unlinked shader.  During
+finalization, metadata entries that depend only on the pipeline state and not on
+the actual shader code can be set.  The `depthClipEnabled` field is one example.
+This entry is set by copying the value of the corresponding entry in the
+pipeline state.  The actual shader code does not matter.  These entries must not
+be present or
+set to 0 in unlinked shaders.  This removes the dependency of unlinked shaders on
+those parts of the pipeline state.
+
+## Epilogues and prologues
+
+Some changes to the text section are too large to be handled by relocations.
+The two main examples are the vertex attribute input formates and the color
+export formates.  These are handled by generating a prologue or epilogue for the
+unlinked shaders.  The prologue and epilogue are also referred to as glue
+shaders, and are implemented by extending the `GlueShader` class in
+[GlueShader.h](../lgc/elfLinker/GlueShader.h).
+
+When generating an unlinked vertex shader, the entry point will have the name
+`_amdgpu_vs_main_fetchless`.  The inputs to this shader will be all of the
+regular inputs to a vertex shader plus all of the vertex input attributes in
+registers.  The PAL metadata will contain a table called `.vertexInputs` that
+will pass along information from the original llvm-ir that would otherwise be
+lost, but is needed to load the vertex inputs correctly.  When linking, a
+prologue is prepended to the unlinked vertex shader that uses the pipeline
+state and the `.vertexInputs` metadata to load all of the vertex input
+attributes into registers. See the `FetchShader` class in
+[GlueShader.cpp](../lgc/elfLinker/GlueShader.cpp)
+for the generation of the prologue. To see how the unlinked shader handles
+vertex inputs see the `LowerVertexFetch` pass in
+[VertexFetch.cpp](../lgc/patch/VertexFetch.cpp).
+
+Fragment shaders will require a prologue to export its outputs.  The unlinked fragment
+shader will have the standard name for the fragment shader.  Instead of
+exporting the outputs, they are left in registers at the end of the shader,
+where there is no `s_endpgm`. Extra information is add to the PAL metadata in
+the `.colorExport` table.  During linking this table and the pipeline state are
+used to generate all of the color export instructions and they are appended to
+the end of the unlinked shader. See the `ColorExportShader` class in
+[GlueShader.cpp](../lgc/elfLinker/GlueShader.cpp).
```

Comment:
nit: extra period in middle of the sentence

[GlueShader.cpp](../lgc/elfLinker/GlueShader.cpp). <----
for the generation of the epilogue.